### PR TITLE
feat: oidc4vp initiate transaction.

### PR DESCRIPTION
**Pull request by vkubiv**
Signed-off-by: Volodymyr Kubiv <volodymyr.kubiv@euristiq.com>

### CODE REVIEW

File: pkg/service/oidc4vp/oidc4vp_service.go [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+}
+
+type RequestObjectClaims struct {
+	VPToken VPToken `json:"vp_token"`
+}
+type VPToken struct {
+	PresentationDefinition *presexch.PresentationDefinition `json:"presentation_definition"`
+}
+
+type RequestObject struct {
```
Comment (aholovko) [inline pkg/service/oidc4vp/oidc4vp_service.go:58]: can we be more specific about what exactly `RequestObject` represents? is it `CredentialAuthorizationRequest`? `PresentationRequest`?

File: pkg/service/oidc4vp/oidc4vp_service.go [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+	ResponseType string              `json:"response_type"`
+	ResponseMode string              `json:"response_mode"`
+	Scope        string              `json:"scope"`
+	Nonce        string              `json:"nonce"`
+	ClientID     string              `json:"client_id"`
+	RedirectURI  string              `json:"redirect_uri"`
+	State        string              `json:"state"`
+	Exp          int64               `json:"exp"`
+	Registration RORegistration      `json:"registration"`
+	Claims       RequestObjectClaims `json:"claims"`
```
Comment (aholovko) [inline pkg/service/oidc4vp/oidc4vp_service.go:70]: plz be consistent in naming: `RORegistration` vs. `RequestObjectClaims`: either `RequestObjectRegistration` and `RequestObjectClaims ` or `RORegistration` and `ROClaims` (I would probably omit RO entirely or use more descriptive name like VPRegistration or whatever that object represents)

File: pkg/service/oidc4vp/txmanager.go [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+	txID, valid, err := tm.nonceStore.GetAndDelete(nonce)
+	if err != nil {
+		return nil, false, fmt.Errorf("oidc tx nonceStore get failed: %w", err)
+	}
+
+	var tx *Transaction
+	if valid {
+		tx, err = tm.txStore.Get(txID)
+		if err != nil {
+			return nil, false, fmt.Errorf("oidc tx get ny  failed: %w", err)
```
Comment (aholovko) [inline pkg/service/oidc4vp/txmanager.go:101]: "oidc tx get ny  failed"?

File: component/profile/reader/file/reader.go [hunk 4]
```diff
@@ -51,10 +54,11 @@ type issuerProfile struct {
...
 type verifierProfile struct {
-	Data                *profileapi.Verifier `json:"verifier,omitempty"`
-	CreateDID           bool                 `json:"createDID"`
-	DidDomain           string               `json:"didDomain"`
-	DidServiceAuthToken string               `json:"didServiceAuthToken"`
+	Data                    *profileapi.Verifier               `json:"verifier,omitempty"`
+	CreateDID               bool                               `json:"createDID"`
+	DidDomain               string                             `json:"didDomain"`
+	DidServiceAuthToken     string                             `json:"didServiceAuthToken"`
+	PresentationDefinitions []*presexch.PresentationDefinition `json:"presentationDefinitions"`
```
Comment (fqutishat) [inline component/profile/reader/file/reader.go:61]: please move it inside profileapi.Verifier

File: component/profile/reader/file/reader.go [hunk 5]
```diff
@@ -178,6 +186,23 @@ func (p *VerifierReader) GetProfile(profileID profileapi.ID) (*profileapi.Verifi
 	return p.verifiers[profileID], nil
 }
 
+func (p *VerifierReader) FindPresentationDefinition(profile *profileapi.Verifier,
```
Comment (fqutishat) [inline component/profile/reader/file/reader.go:189]: remove it when you move PresentationDefinition inside rofileapi.Verifier

File: docs/v1/openapi.yaml [hunk 6]
```diff
@@ -535,6 +560,25 @@ components:
...
         - expires_in
+    InitiateOIDCData:
+      title: InitiateOIDCData
+      type: object
+      properties:
+        presentation_definition_id:
+          type: string
+        purpose:
+          type: string
+    InitiateOIDCResponse:
```
Comment (aholovko) [inline docs/v1/openapi.yaml:571]: plz be more specific in naming those models as well - Initiate OIDC Issuance also has similar structures

Comment (vkubiv) [reply docs/v1/openapi.yaml:571]: yes yes

File: pkg/service/oidc4vp/oidc4vp_service.go (hunk 1 above)
Comment (vkubiv) [reply pkg/service/oidc4vp/oidc4vp_service.go:60]: Comment is added

**Review by aholovko — APPROVED** [review]

File: component/profile/reader/file/reader.go [hunk 7]
```diff
@@ -34,8 +36,9 @@ type IssuerReader struct {
 
 // VerifierReader read verifier profiles.
 type VerifierReader struct {
-	issuers   map[string]*profileapi.Issuer
-	verifiers map[string]*profileapi.Verifier
+	issuers     map[string]*profileapi.Issuer
+	verifiers   map[string]*profileapi.Verifier
+	verifierPDs map[string][]*presexch.PresentationDefinition
```
[the remark this answers is not in the document]
Comment (fqutishat) [reply component/profile/reader/file/reader.go:41]: call it presentationDefinition

File: pkg/profile/api.go [hunk 8]
```diff
@@ -52,15 +53,16 @@ type VCConfig struct {
...
+	ID                      ID                                 `json:"id,omitempty"`
+	Name                    string                             `json:"name,omitempty"`
+	URL                     string                             `json:"url,omitempty"`
+	Active                  bool                               `json:"active,omitempty"`
+	OrganizationID          string                             `json:"organizationID,omitempty"`
+	Checks                  *VerificationChecks                `json:"checks,omitempty"`
+	OIDCConfig              *OIDC4VPConfig                     `json:"oidcConfig,omitempty"`
+	KMSConfig               *vcskms.Config                     `json:"kmsConfig,omitempty"`
+	SigningDID              *SigningDID                        `json:"signingDID,omitempty"`
+	PresentationDefinitions []*presexch.PresentationDefinition `json:"presentationDefinitions"`
```
Comment (fqutishat) [inline pkg/profile/api.go:65]: add omitempty

**Review by fqutishat — APPROVED** [review]

**Pull request merged**